### PR TITLE
feat: extract indexer/delegators' reward by era

### PR DESCRIPTION
**Pull request by icezohu**

### CODE REVIEW

File: src/mappings/rewardsDistributor.ts [hunk 1]
```diff
@@ -40,7 +42,7 @@ export async function handleRewardsDistributed(
   );
   assert(event.args, 'No event args');
 
-  const { indexer, eraIdx } = event.args;
+  const { indexer, eraIdx, rewards: eventRewards } = event.args;
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: You can get commission from this event, which you will need to create commission part of EraReward

File: src/mappings/rewardsDistributor.ts [hunk 2]
```diff
@@ -90,7 +94,33 @@ export async function handleRewardsDistributed(
...
     }
+
+    if (rewards.gt(0)) {
+      const eraReward = await updateEraReward({
+        indexerId: indexer,
+        delegatorId: delegator.delegatorId,
+        eraId: eraIdx.toHexString(),
+        isCommission: false,
+        claimed: false,
+        amount: rewards.toBigInt() - rewardOld,
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: The reward from this event is the reward of this era, so you don't need to minus it by rewardOld
```suggestion
        amount: rewards.toBigInt(),
```

File: src/mappings/rewardsDistributor.ts [hunk 3]
```diff
@@ -90,7 +94,33 @@ export async function handleRewardsDistributed(
...
   }
+
+  await updateEraReward({
+    indexerId: indexer,
+    delegatorId: indexer,
+    eraId: eraIdx.toHexString(),
+    isCommission: true,
+    claimed: true, // commission rewards already in indexer's account
+    // amount: rewards.mul(commission).div(100).toBigInt(),
+    amount: eventRewards.toBigInt() - accumulatedEraRewards,
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: commission should be get from event.args

File: src/mappings/rewardsDistributor.ts [hunk 4]
```diff
@@ -129,6 +159,90 @@ export async function handleRewardsClaimed(
...
+}
+
+async function updateEraReward(data: EraRewardData): Promise<EraReward> {
+  console.log('updateEraReward', data);
+
+  const id = `${data.indexerId}_${data.delegatorId}_${data.eraId}${
+    data.isCommission ? '_commission' : ''
+  }`;
+
+  let eraReward = await EraReward.get(id);
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: You should know if you want to create or update `EraReward`, you can gain some performance by skipping this step.
`Update` only happens in claim rewards.

File: src/mappings/rewardsDistributor.ts [hunk 5]
```diff
@@ -90,7 +94,33 @@ export async function handleRewardsDistributed(
...
+        claimed: false,
+        amount: rewards.toBigInt() - rewardOld,
+        createdBlock: event.blockNumber,
+        createdTimestamp: biToDate(event.block.timestamp),
+      });
+      accumulatedEraRewards += eraReward.amount;
+    }
   }
+
+  await updateEraReward({
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: if `rewards` can be zero, then commission also can be. You probably will want to skip that as well.

File: src/mappings/rewardsDistributor.ts [hunk 6]
```diff
@@ -129,6 +159,90 @@ export async function handleRewardsClaimed(
...
+      id,
+      lastClaimedEra: 0,
+    });
+  }
+
+  let lastClaimedEra = eraRewardClaimed.lastClaimedEra;
+  const hasNextEraReward = true;
+
+  while (hasNextEraReward) {
+    const eraReward = await EraReward.get(`${id}_${lastClaimedEra + 1}`);
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: Need to compare with current era id. only era up to `CurrentEra-1` can be marked as claimed.

File: src/mappings/staking.ts [hunk 7]
```diff
@@ -139,6 +146,7 @@ export async function handleAddDelegation(
   await delegation.save();
   await updateIndexerCapacity(indexer, event);
   await updateMaxUnstakeAmount(indexer, event);
+  await updateStakeSummary(event, eraManager);
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: make sure currentEra is only queried once in each handler, and can even provide cross handler cache for it via utilising global variable

File: src/mappings/staking.ts [hunk 8]
```diff
@@ -278,3 +286,78 @@ export async function handleWithdrawCancelled(
...
+  }
+
+  const eraIdx = await getCurrentEra(eraManager);
+  const eraId = eraIdx.toString();
+  // const prevEraIdx = eraIdx - 1;
+  const prevEraId = (eraIdx - 1).toString();
+  const nextEraIdx = eraIdx + 1;
+  const nextEraId = nextEraIdx.toString();
+
+  const isFirstStake = !!(await StakeSummary.get(prevEraId));
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: That's not enough, indexer can add stakes multiple times in currentEra and only the first one is considered first stake.

Comment (ianhe8x) [reply src/mappings/staking.ts]: Maybe use `!(await StakeSummary.get(prevEraId)) && !(await StakeSummary.get(currentEraId))

File: src/mappings/staking.ts [hunk 9]
```diff
@@ -278,3 +286,78 @@ export async function handleWithdrawCancelled(
...
+  } else {
+    delegatorStake += amountBn;
+  }
+
+  const eraIdx = await getCurrentEra(eraManager);
+  const eraId = eraIdx.toString();
+  // const prevEraIdx = eraIdx - 1;
+  const prevEraId = (eraIdx - 1).toString();
+  const nextEraIdx = eraIdx + 1;
+  const nextEraId = nextEraIdx.toString();
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: I think currently eraId is hex form of era number.

File: src/mappings/staking.ts [hunk 10]
```diff
@@ -278,3 +286,78 @@ export async function handleWithdrawCancelled(
...
+  if (source === indexer) {
+    indexerStake += amountBn;
+  } else {
+    delegatorStake += amountBn;
+  }
+
+  const eraIdx = await getCurrentEra(eraManager);
+  const eraId = eraIdx.toString();
+  // const prevEraIdx = eraIdx - 1;
+  const prevEraId = (eraIdx - 1).toString();
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: same

File: src/mappings/staking.ts [hunk 11]
```diff
@@ -278,3 +286,78 @@ export async function handleWithdrawCancelled(
...
+  let delegatorStake = BigInt(0);
+
+  if (source === indexer) {
+    indexerStake += amountBn;
+  } else {
+    delegatorStake += amountBn;
+  }
+
+  const eraIdx = await getCurrentEra(eraManager);
+  const eraId = eraIdx.toString();
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: same

File: src/mappings/staking.ts [hunk 12]
```diff
@@ -278,3 +286,78 @@ export async function handleWithdrawCancelled(
...
+  // const prevEraIdx = eraIdx - 1;
+  const prevEraId = (eraIdx - 1).toString();
+  const nextEraIdx = eraIdx + 1;
+  const nextEraId = nextEraIdx.toString();
+
+  const isFirstStake = !!(await StakeSummary.get(prevEraId));
+
+  if (isFirstStake) {
+    let currentStakeSummary = await StakeSummary.get(eraId);
+    if (!currentStakeSummary) {
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: when isFirstStake, StakeSummary shouldn't exist, so you can skip the check

File: src/mappings/staking.ts [hunk 13]
```diff
@@ -278,3 +286,78 @@ export async function handleWithdrawCancelled(
...
+  const indexerEntity = await Indexer.get(indexer);
+  assert(indexerEntity, `Indexer ${indexer} does not exist`);
+  // const isFirstStake =
+  //   BigInt.fromJSONType(indexerEntity.totalStake.value) === BigInt(0);
+
+  let indexerStake = BigInt(0);
+  let delegatorStake = BigInt(0);
+
+  if (source === indexer) {
+    indexerStake += amountBn;
```
Comment (ianhe8x) [inline src/mappings/staking.ts]:
```suggestion
    indexerStake = amountBn;
```

File: src/mappings/staking.ts [hunk 14]
```diff
@@ -278,3 +286,78 @@ export async function handleWithdrawCancelled(
...
+  assert(event.args, 'No event args');
+  const { source, indexer, amount } = event.args;
+  const amountBn = amount.toBigInt();
+
+  const indexerEntity = await Indexer.get(indexer);
+  assert(indexerEntity, `Indexer ${indexer} does not exist`);
+  // const isFirstStake =
+  //   BigInt.fromJSONType(indexerEntity.totalStake.value) === BigInt(0);
+
+  let indexerStake = BigInt(0);
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: for better readibility, and same to delegatorStake
```suggestion
  let newIndexerStake = BigInt(0);
```

File: src/mappings/staking.ts (hunk 7 above)
Comment (icezohu) [reply src/mappings/staking.ts]: should I refactor the 5 functions above to meet this requirement? it's been 5 same queries before this query...

File: src/mappings/rewardsDistributor.ts (hunk 4 above)
Comment (icezohu) [reply src/mappings/rewardsDistributor.ts]: should I consider if it's doing regenerating database from the chain and validate before creating the records?

File: src/mappings/staking.ts (hunk 8 above)
Comment (icezohu) [reply src/mappings/staking.ts]: I need to know if it's only one senario that should create current stake summary -- no previous summary and no current summary? which means only one stake amount will be added to `first stake summary`, no more.

File: src/mappings/staking.ts (hunk 7 above)
Comment (ianhe8x) [reply src/mappings/staking.ts]: Yes please

File: src/mappings/staking.ts (hunk 8 above)
Comment (ianhe8x) [reply src/mappings/staking.ts]: check here https://github.com/subquery/network-contracts/blob/f3ebee46ed92b044e52d744f4017ef5c47884a2f/contracts/Staking.sol#L267

File: src/mappings/rewardsDistributor.ts (hunk 4 above)
Comment (ianhe8x) [reply src/mappings/rewardsDistributor.ts]: reindex will always start from empty database

File: src/mappings/rewardsDistributor.ts [hunk 15]
```diff
@@ -195,17 +192,13 @@ async function updateEraReward(data: EraRewardData): Promise<EraReward> {
...
-    eraReward.claimed = data.claimed;
 
-    eraReward.amount = data.amount ?? eraReward.amount;
-    eraReward.createdBlock = data.createdBlock ?? eraReward.createdBlock;
-    eraReward.createdTimestamp =
-      data.createdTimestamp ?? eraReward.createdTimestamp;
+    await eraReward.save();
+    return eraReward;
+  } catch (e) {
+    logger.error('ERROR: createEraReward', e);
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: pls re-throw the error here

File: src/mappings/rewardsDistributor.ts [hunk 16]
```diff
@@ -195,17 +192,13 @@ async function updateEraReward(data: EraRewardData): Promise<EraReward> {
...
   }
-
-  await eraReward.save();
-  return eraReward;
+  return null;
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: not supposed to happen

File: src/mappings/rewardsDistributor.ts [hunk 17]
```diff
@@ -217,31 +210,35 @@ async function updateEraRewardClaimed(
...
 
-  if (lastClaimedEra > eraRewardClaimed.lastClaimedEra) {
-    eraRewardClaimed.lastClaimedEra = lastClaimedEra;
-    await eraRewardClaimed.save();
+    if (lastClaimedEra > eraRewardClaimed.lastClaimedEra) {
+      eraRewardClaimed.lastClaimedEra = lastClaimedEra;
+      await eraRewardClaimed.save();
+    }
+  } catch (e) {
+    logger.error('ERROR: updateEraRewardClaimed', e);
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: re-throw the error

File: src/mappings/rewardsDistributor.ts [hunk 18]
```diff
@@ -217,31 +210,35 @@ async function updateEraRewardClaimed(
...
-  const hasNextEraReward = true;
+    const currentEra = await getCurrentEra();
+    let lastClaimedEra = eraRewardClaimed.lastClaimedEra;
 
-  while (hasNextEraReward) {
-    const eraReward = await EraReward.get(`${id}_${lastClaimedEra + 1}`);
-    if (!eraReward) break;
+    while (lastClaimedEra + 1 < currentEra) {
+      const eraReward = await EraReward.get(`${id}_${lastClaimedEra + 1}`);
+      if (!eraReward) break;
```
Comment (ianhe8x) [inline src/mappings/rewardsDistributor.ts]: I feel you should use `continue` here

File: src/mappings/staking.ts [hunk 19]
```diff
@@ -288,76 +259,61 @@ export async function handleWithdrawCancelled(
...
-      indexerStake,
-      delegatorStake,
-    });
-  } else {
-    nextStakeSummary.totalStake += amountBn;
-    nextStakeSummary.indexerStake += indexerStake;
-    nextStakeSummary.delegatorStake += delegatorStake;
+    await nextStakeSummary.save();
+  } catch (e) {
+    logger.error('Error: updateStakeSummary', e);
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: rethrow error here

File: src/mappings/staking.ts [hunk 20]
```diff
@@ -278,42 +280,93 @@ async function updateStakeSummary(
...
+
+    // update IndexerStakeSummary
+
+    let indexerStakeSummary = await IndexerStakeSummary.get(indexer);
+    let isFirstStake = false;
+
+    if (!indexerStakeSummary) {
+      isFirstStake = true;
+    } else if (
+      indexerStakeSummary.eraId === currEraId &&
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: for better readability, you can extract a getter function for totalStake
`getTotalStake(summary: IndexerStakeSummary, currentEra: number): BigInt`

File: src/mappings/staking.ts [hunk 21]
```diff
@@ -278,42 +280,93 @@ async function updateStakeSummary(
...
-      const currentStakeSummary = StakeSummary.create({
-        id: currEraId,
+    if (!indexerStakeSummary) {
+      indexerStakeSummary = IndexerStakeSummary.create({
+        id: indexer,
+        eraId: currEraId,
         totalStake: amountBn,
         indexerStake: newIndexerStake,
         delegatorStake: newDelegatorStake,
+        nextTotalStake: BigInt(0),
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: next*Stake should be equal to current stake's value
I see you have the logic of increasing next*Stake out of the if tree, though maybe move that logic into if blocks is clearer

File: src/mappings/staking.ts [hunk 22]
```diff
@@ -278,42 +280,93 @@ async function updateStakeSummary(
...
+        nextIndexerStake: BigInt(0),
+        nextDelegatorStake: BigInt(0),
       });
-      await currentStakeSummary.save();
+    } else if (isFirstStake) {
+      indexerStakeSummary.totalStake = amountBn;
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: doesn't seem to be any different from the above block of code

File: src/mappings/staking.ts [hunk 23]
```diff
@@ -278,42 +280,93 @@ async function updateStakeSummary(
...
+      indexerStakeSummary.indexerStake = indexerStakeSummary.nextIndexerStake;
+      indexerStakeSummary.delegatorStake =
+        indexerStakeSummary.nextDelegatorStake;
     }
 
-    let nextStakeSummary = await StakeSummary.get(nextEraId);
-    if (!nextStakeSummary) {
-      nextStakeSummary = StakeSummary.create({
-        id: nextEraId,
+    indexerStakeSummary.eraId = currEraId;
```
Comment (ianhe8x) [inline src/mappings/staking.ts:414]: move `indexerStakeSummary.eraId = currEraId;` into `if (indexerStakeSummary.eraId !== currEraId)` block

File: src/mappings/staking.ts [hunk 24]
```diff
@@ -278,42 +280,93 @@ async function updateStakeSummary(
...
+    indexerStakeSummary.nextTotalStake += amountBn;
+    indexerStakeSummary.nextIndexerStake += newIndexerStake;
+    indexerStakeSummary.nextDelegatorStake += newDelegatorStake;
+    await indexerStakeSummary.save();
+
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: You forget to update `IndexerStakeSummary` for all indexers. 
You can pick an unique id to represent the global stake summary.

File: src/mappings/staking.ts (hunk 22 above)
Comment (icezohu) [reply src/mappings/staking.ts]: this block is for first stake and summary already exists

File: src/mappings/staking.ts [hunk 25]
```diff
@@ -278,42 +280,93 @@ async function updateStakeSummary(
...
+        eraId: currEraId,
         totalStake: amountBn,
         indexerStake: newIndexerStake,
         delegatorStake: newDelegatorStake,
       });
+      await currIndexerStake.save();
+    }
+
+    let nextIndexerStake = await IndexerStake.get(nextIndexerStakeId);
+    if (!nextIndexerStake) {
```
Comment (ianhe8x) [inline src/mappings/staking.ts]: `.save()` in subql is always upsert, so you don't have to check the existence of  `IndexerStake.get(nextIndexerStakeId)`

File: src/mappings/staking.ts (hunk 21 above)
Comment (icezohu) [reply src/mappings/staking.ts]: ok, I feel the logic is quite awk in my way. I will try to make it clearer.

**Comment by mzxyz** [comment]
@icezohu got the following error with the latest changes:
```
{"type":"error","name":"SequelizeUnknownConstraintError","message":"Unknown constraint error","stack":"Error\n    
at Query.run (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-
sequelize/lib/dialects/postgres/query.js:51:25)\n    at /usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-sequelize/lib/sequelize.js:315:28\n    at process.processTicksAndRejections 
(node:internal/process/task_queues:95:5)\n    at async PostgresQueryInterface.addIndex 
(/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-sequelize/lib/dialects/abstract/query-
interface.js:250:12)\n    at async EraReward.sync (/usr/local/lib/node_modules/@subql/node-
ethereum/node_modules/@subql/x-sequelize/lib/model.js:999:7)\n    at async Sequelize.sync 
(/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-sequelize/lib/sequelize.js:377:9)\n    at async 
StoreService.syncSchema (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/node-
core/dist/indexer/store.service.js:290:9)\n    at async StoreService.init (/usr/local/lib/node_modules/@subql/node-
ethereum/node_modules/@subql/node-core/dist/indexer/store.service.js:81:13)\n    at async initDbSchema 
(/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/node-core/dist/utils/project.js:197:5)\n    at 
async ProjectService.initDbSchema (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/node-
core/dist/indexer/project.service.js:120:9)"},"message":"Having a problem when syncing schema"}

```

**Review by mzxyz — APPROVED** [review]

**Comment by icezohu** [comment]
> @icezohu got the following error with the latest changes:
> 
> ```
> {"type":"error","name":"SequelizeUnknownConstraintError","message":"Unknown constraint error","stack":"Error\n    
> at Query.run (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-
> sequelize/lib/dialects/postgres/query.js:51:25)\n    at /usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-sequelize/lib/sequelize.js:315:28\n    at process.processTicksAndRejections 
> (node:internal/process/task_queues:95:5)\n    at async PostgresQueryInterface.addIndex 
> (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-sequelize/lib/dialects/abstract/query-
> interface.js:250:12)\n    at async EraReward.sync (/usr/local/lib/node_modules/@subql/node-
> ethereum/node_modules/@subql/x-sequelize/lib/model.js:999:7)\n    at async Sequelize.sync 
> (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/x-sequelize/lib/sequelize.js:377:9)\n    at async 
> StoreService.syncSchema (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/node-
> core/dist/indexer/store.service.js:290:9)\n    at async StoreService.init (/usr/local/lib/node_modules/@subql/node-
> ethereum/node_modules/@subql/node-core/dist/indexer/store.service.js:81:13)\n    at async initDbSchema 
> (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/node-core/dist/utils/project.js:197:5)\n    at 
> async ProjectService.initDbSchema (/usr/local/lib/node_modules/@subql/node-ethereum/node_modules/@subql/node-
> core/dist/indexer/project.service.js:120:9)"},"message":"Having a problem when syncing schema"}
> ```

I set @index on two boolean fields which made this error. It's fixed and validated now.

**Pull request merged**